crates/ironrdp-rdpdr-native/src/nix/simple.rs

```diff
@@ -0,0 +1,778 @@
+                                device_io_reply: DeviceIoResponse::new(
+                                    req_inner.device_io_request,
+                                    NtStatus::UNSUCCESSFUL,
+                                ),
+                                buffer: None,
+                            },
+                        ),
+                    )])
+                }
+            } else {
+                warn!("no such file");
+                let res = RdpdrPdu::ClientDriveQueryInformationResponse(ClientDriveQueryInformationResponse {
+                    device_io_response: DeviceIoResponse::new(req_inner.device_io_request, NtStatus::NO_SUCH_FILE),
+                    buffer: None,
+                });
+                Ok(vec![SvcMessage::from(res)])
+            }
+        }
+        None => {
+            warn!("no such file");
+            let res = RdpdrPdu::ClientDriveQueryInformationResponse(ClientDriveQueryInformationResponse {
+                device_io_response: DeviceIoResponse::new(req_inner.device_io_request, NtStatus::NO_SUCH_FILE),
+                buffer: None,
+            });
+            Ok(vec![SvcMessage::from(res)])
+        }
+    }
+}
+
+pub(crate) fn set_information(
+    backend: &mut SimpleRdpdrBackend,
+    req_inner: ServerDriveSetInformationRequest,
+) -> PduResult<Vec<SvcMessage>> {
+    match backend.file_path_map.get(&req_inner.device_io_request.file_id) {
+        Some(file) => {
+            match &req_inner.set_buffer {
+                FileInformationClass::Rename(info) => {
+                    let mut to = backend.file_base.clone();
+                    to.push_str(&info.file_name.replace('\\', "/"));
+                    if let Err(error) = std::fs::rename(file, to) {
+                        warn!(?error, "Rename file error");
+                        let res = RdpdrPdu::ClientDriveSetInformationResponse(ClientDriveSetInformationResponse::new(
+                            &req_inner,
+                            NtStatus::UNSUCCESSFUL,
+                        )?);
+                        return Ok(vec![SvcMessage::from(res)]);
+                    }
+                }
+                FileInformationClass::Allocation(_) => {
+                    //nothing to do
+                }
+                FileInformationClass::Disposition(_) => {
+                    if let Err(error) = std::fs::remove_file(file) {
+                        warn!(?error, "Remove file error");
+                        let res = RdpdrPdu::ClientDriveSetInformationResponse(ClientDriveSetInformationResponse::new(
+                            &req_inner,
+                            NtStatus::UNSUCCESSFUL,
+                        )?);
+                        return Ok(vec![SvcMessage::from(res)]);
+                    }
+                }
+                FileInformationClass::EndOfFile(info) => {
+                    let mut set_end_res = -1;
+                    // SAFETY: it is safe to call on *nix platform
```

Comment:
question: This is not really documenting why the unsafe operation is indeed safe. Per the documentation, it would be something like that:

```suggestion
                    // SAFETY: `file` is a regular file that we opened with write access.
```

However, I see `file_map` may contain files with `read` permission only. (See line 699.)
Can you explain how we are never calling `ftruncate` on these?

Reply:
ok, i got it